src/runtime_src/core/pcie/tools/xbutil/xbutil.cpp

```diff
@@ -2019,3 +2032,182 @@ int xcldev::device::testM2m()
+        throw std::runtime_error("Cannot obtain output BO");
+    }
+
+    memset(info->output_bo_ptr, 'o', info->out_size);
+    if(xclSyncBO(handle, info->out_bo_handle, XCL_BO_SYNC_BO_TO_DEVICE, info->out_size, 0)) {
+        iops_free_unmap_bo(handle, info->out_bo_handle, info->output_bo_ptr, info->out_size);
+        throw std::runtime_error("Cannot sync output BO to device");
+    }
+
+    xclBOProperties prop;
+    if(xclGetBOProperties(handle, info->out_bo_handle, &prop)) {
+        iops_free_unmap_bo(handle, info->exec_bo_handle, info->exec_bo_ptr, info->exec_size);
+        throw std::runtime_error("Cannot obtain BO dev address");
+    }
+    uint64_t boh_address = prop.paddr;
+
+    info->exec_bo_handle = xclAllocBO(handle, info->exec_size, 0, XCL_BO_FLAGS_EXECBUF);
+    info->exec_bo_ptr = reinterpret_cast<ert_start_kernel_cmd *>(xclMapBO(handle, info->exec_bo_handle, true));
+    if (info->exec_bo_ptr == nullptr) {
+        iops_free_unmap_bo(handle, info->exec_bo_handle, info->exec_bo_ptr, info->exec_size);
+        throw std::runtime_error("Cannot obtain exec buf BO");
+    }
+    std::memset(info->exec_bo_ptr, 0, info->exec_size);
+
+    //construct the exec buffer cmd to start the kernel.
+    int rsz = 19; // regmap array size
+    info->exec_bo_ptr->state = 0;
+    info->exec_bo_ptr->opcode = ERT_START_CU;
+    info->exec_bo_ptr->count = rsz;
+    info->exec_bo_ptr->cu_mask = (0x1 << 0);
+    info->exec_bo_ptr->data[rsz - 3] = boh_address;
+    info->exec_bo_ptr->data[rsz - 2] = (boh_address >> 32);
+}
+
+static void
+execute_cmds(xclDeviceHandle handle, std::vector<std::shared_ptr<exec_struct>>& cmds, 
+    unsigned int num_execs, double& duration)
+{
+    unsigned int num_issued = 0;
+    unsigned int num_completed = 0;
+    auto start = std::chrono::high_resolution_clock::now();
+    for (auto& c : cmds) {
+        if(xclExecBuf(handle, c->exec_bo_handle))
+            throw std::runtime_error("Unable to issue exec buf");
+        num_issued++;
+    }
+    while (num_completed < num_execs) {
+        for (auto& c : cmds) {
+            // c->wait();
+            while (c->exec_bo_ptr->state < ERT_CMD_STATE_COMPLETED) {
+                while (xclExecWait(handle, -1) == 0);
+            }
+            if(c->exec_bo_ptr->state != ERT_CMD_STATE_COMPLETED)
+                throw std::runtime_error("CU execution failed");
+                //wait end
+            num_completed++;
+            if (num_completed >= num_execs)
+                break;
+            if (num_issued < num_execs) {
+                if(xclExecBuf(handle, c->exec_bo_handle))
+                    throw std::runtime_error("Unable to issue exec buf");
+                num_issued++;
+            }
+        }
+    }
+    auto end = std::chrono::high_resolution_clock::now();
+    duration = (std::chrono::duration_cast<std::chrono::microseconds>
+    (end - start)).count();
+}
+
+int 
+xcldev::device::iopsTest()
```

Comment:
This function is too long.  For example the triple for-loop starting at line 2177 has to be broken.